### PR TITLE
Audit 12: Improve module upgrade flow

### PR DESCRIPTION
**Pull request by adairrr**
This change addresses issue 12 on our audit, focusing on module upgrades via the manager. 

It implements the following changes:
- Verify that there are updates in the migrations (non-empty targets)
- Verify that there are no duplicate migrations being performed
- Migrate manager last in case of migration with other modules.
- Refactor migration logic for readability

On the refactoring front, the changes were to extract the `api`, `app`, and `standalone` migration logic into their own functions. The logic within them were not changed. The common functionality between them was adding the migrate msgs for each module to the context, which was extracted into `add_module_upgrade_to_context`. 

Additionally, to account for the manager module being migrated after each other module, its info is extracted into a tuple, then the migrate message is built and appended to the submessages. There is no context set for the manager migration, as it happens after each other module is migrated.

### CODE REVIEW

**Comment by adairrr** [comment]
Task linked: [CU-861mkcftz Improve input validation when upgrading modules](https://app.clickup.com/t/861mkcftz)

**Comment by CyberHoward** [comment]
Would be really nice if you could give a recap on any architectural changes you made because the migration logic is kinda intricate so I want to be sure that i'm reviewing that explicitly. Right now I don't know if you made any of those changes and if you did, no idea where to start.

File: contracts/account/manager/src/commands.rs [hunk 1]
```diff
@@ -140,8 +140,7 @@ pub fn register_module(
             // assert version requirements
             let dependencies = versioning::assert_install_requirements(deps.as_ref(), &id)?;
             versioning::set_as_dependent(deps.storage, id, dependencies)?;
-            response = response.add_message(allowlist_dapp_on_proxy(
-                deps.as_ref(),
+            response = response.add_message(add_module_from_proxy(
```
[the remark this answers is not in the document]
Comment (adairrr) [reply contracts/account/manager/src/commands.rs:143]: Good catch. Should've been `to`

**Comment by CyberHoward** [comment]
> Would be really nice if you could give a recap on any architectural changes you made because the migration logic is kinda intricate so I want to be sure that i'm reviewing that explicitly. Right now I don't know if you made any of those changes and if you did, no idea where to start.

So no architectural changes?

File: contracts/account/manager/src/commands.rs [hunk 2]
```diff
@@ -291,72 +308,119 @@ pub fn upgrade_modules(
...
+/// Add the module upgrade to the migration context and check for duplicates
+fn add_module_upgrade_to_context(
+    storage: &mut dyn Storage,
+    module_id: &str,
+    module_deps: Vec<Dependency>,
+) -> Result<(), ManagerError> {
+    // Add module upgrade to reply context
+    let update_context = |mut upgraded_modules: Vec<(String, Vec<Dependency>)>| -> StdResult<Vec<(String, Vec<Dependency>)>> {
+        if upgraded_modules.iter().any(|(id, _)| id == module_id) {
+            Err(StdError::generic_err("Module already upgraded.".to_string()))
```
Comment (CyberHoward) [inline contracts/account/manager/src/commands.rs:407]: Is this check not better performed before any messages/state is changed? Right when we parse the message we can check if there are duplicate migration requests. Should be cheaper gas wise.

**Review by CyberHoward — CHANGES_REQUESTED** [review]
little comment, should be good to merge after!

File: contracts/account/manager/src/commands.rs (hunk 2 above)
Comment (adairrr) [reply contracts/account/manager/src/commands.rs:407]: Nice catch!

**Comment by adairrr** [comment]
> So no architectural changes?

I updated the description to include the changes that were made. The logic remains the same.

**Review by CyberHoward — APPROVED** [review]

**Pull request merged**